mapbox/node-pre-gyp
# Refactor Mock and Test Setup

### Overview

This pull request refactors the S3 mocking setup (see: #540) simplifying implementation, separates the S3 tests from the build tests, introduces test apps using alternative bucket specification (see: #576) and production/staging hosts (see: #533) and enables easy switch between mock bucket and real bucket (given status of: #613).

This pull request comes "on top" of (i.e. includes changes from) #648, #649, #650.

### Change
#### Code
* Refactored mock setup (e3c9772)
    - mock modules moved to `lib/mock`.
    - mock modules are self contained and have a signature similar to that of other modules in the package.
    - simplified http mock to remove no longer needed interface.
    - modified http mock to support path style bucket access.
    - set http mock to work with default `npg-mock-bucket` bucket.
* Refactored test setup (9085b88)
    - separated s3 tests from build tests.
    - refactored fetch test and proxy-bcrypt test to work with refactored mock setup.
* Refactored test apps and increased test coverage (4572c86)
    - set hosts of test apps to point to mock bucket.
    - added app1.1 - identical to app1 but using production and staging binary host option.
    - added app1.2 - identical to app1 but using explicit host, region, bucket options.
* Added service script to switch test apps bucket from mock to S3 and vis versa. Removed previous one (c2c54c5)

#### Testing
- "Out of the Box" (repo clone) tests will skip interaction with S3. 
- Setting the environment variable `node_pre_gyp_mock_s3` to true will run them against the mock ([see: repo CI](https://app.travis-ci.com/github/mapbox/node-pre-gyp/jobs/569765077)).
- To run against a real bucket:
    - unset environment variable `node_pre_gyp_mock_s3`.
    - set environment variables `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`.
    - run `npm run bucket {name-of-your-us-east-1-bucket}`
- To switch back to mock:
    - set environment variable `node_pre_gyp_mock_s3` to true.
    - run `npm run bucket`
- Manually configuring regions other than `us-east-1` also works but switch is more work.
- Results of running `npm run coverage` differ depending on what bucket is used (real s3 or mock).

### Tests
- All tests pass in both real bucket and mock bucket setups (including node 18).
